engine/crates/runtime-local/src/rate_limiting/redis.rs

```diff
@@ -207,25 +207,21 @@ impl RedisRateLimiter {
         pipe.atomic();
 
         pipe.cmd("GET").arg(&previous_bucket);
-        pipe.cmd("INCR").arg(&current_bucket);
-
-        // Sets the timeout to the set. This will delete the data after the duration if we do not modify the value.
-        pipe.cmd("EXPIRE")
-            .arg(&current_bucket)
-            .arg(config.duration.as_secs() * 2)
-            .ignore();
+        pipe.cmd("GET").arg(&current_bucket);
 
         // Execute the whole pipeline in one multiplexed request.
-        match pipe.query_async::<_, (Option<u64>, u64)>(&mut *conn).await {
+        match pipe.query_async::<_, (Option<u64>, Option<u64>)>(&mut *conn).await {
             Ok((previous_count, current_count)) => {
                 let previous_count = previous_count.unwrap_or_default().min(config.limit as u64);
-                let current_count = current_count.min(config.limit as u64);
+                let current_count = current_count.unwrap_or_default().min(config.limit as u64);
 
                 // Sum is a percentage what is left from the previous window, and the count of the
                 // current window.
                 let average = previous_count as f64 * (1.0 - bucket_percentage) + current_count as f64;
 
-                if average <= config.limit as f64 {
+                if average < config.limit as f64 {
```

Comment:
It's not obvious that doing two requests is better than everything at once. Have you measured it? 
Cloudflare doesn't make *any* requests to their Memcache if the user is rate-limited. They keep track of those in-memory.

Reply:
We cannot increment this value in a hot loop in retries. And if we don't increment, the <= check would lead to a wrong result. Now if we do it like this, we load the values first, retry as many times as needed and _when_ we have a successful result, we increment the value asynchronously in a task.